### PR TITLE
ref(notifications): Add digest emails to Notifications hierarchy

### PR DESCRIPTION
**Pull request by mgaeta**
This is the last notification type that needs to be integrated into the Notifications hierarchy. Some features I needed to add:
 - pass a `referrer` parameter to the unsubscribe link generator
 - pass `context` into the subject generator (in a previous PR)

### CODE REVIEW

File: src/sentry/digests/utilities.py [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
-    end = None
+def get_digest_metadata(
+    digest: Digest,
+) -> Tuple[Optional[datetime], Optional[datetime], CounterT[str]]:
+    """TODO(mgaeta): This should probably just be part of `build_digest`."""
+    start: Optional[datetime] = None
+    end: Optional[datetime] = None
 
-    counts = Counter()
+    counts: CounterT[str] = Counter()
```
Comment (ceorourke) [inline src/sentry/digests/utilities.py:19]: is this a typo or on purpose?

File: src/sentry/mail/notifications.py [hunk 2]
```diff
@@ -34,28 +34,26 @@ def get_headers(notification: BaseNotification) -> Mapping[str, Any]:
...
+    return f"{prefix}{notification.get_subject(context)}".encode("utf-8")
 
 
-def get_unsubscribe_link(user_id: int, group_id: int) -> str:
+def get_unsubscribe_link(user_id: int, resource_id: int, key: str = "issue") -> str:
     return generate_signed_link(
         user_id,
-        "sentry-account-email-unsubscribe-issue",
-        kwargs={"issue_id": group_id},
+        f"sentry-account-email-unsubscribe-{key}",
```
Comment (ceorourke) [inline src/sentry/mail/notifications.py:52]: what would this be if not an issue?

File: src/sentry/digests/utilities.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/digests/utilities.py:19]: The name `Counter` is already being used by 
```python
from collections import Counter 
```
so I aliased the type:
```python
from typing import Counter as CounterT
```
I'm not totally happy with the `*T` suffix. Would `CounterType` make more sense?

Comment (ceorourke) [reply src/sentry/digests/utilities.py:19]: ooh yeah `CounterType` would be more clear

File: src/sentry/mail/notifications.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/mail/notifications.py:52]: We have another endpoint for unsubscribing from a project
https://github.com/getsentry/sentry/blob/master/src/sentry/mail/adapter.py#L164-L170
https://github.com/getsentry/sentry/blob/master/src/sentry/web/frontend/accounts.py#L226-L252

File: src/sentry/digests/__init__.py [hunk 3]
```diff
@@ -24,6 +25,8 @@ def datetime(self):
 
 OPTIONS = frozenset(("increment_delay", "maximum_delay", "minimum_delay"))
 
+Digest = Mapping[str, Mapping[str, Any]]
```
Comment (ceorourke) [inline src/sentry/digests/__init__.py:28]: I think this is from the digests typing PR

File: tests/sentry/mail/test_adapter.py [hunk 4]
```diff
@@ -216,41 +218,6 @@ def test_get_sendable_user_objects(self):
         assert user4 not in self.adapter.get_sendable_user_objects(project)
 
 
-class MailAdapterBuildSubjectPrefixTest(BaseMailAdapterTest, TestCase):
```
Comment (ceorourke) [inline tests/sentry/mail/test_adapter.py:219]: why just remove these instead of updating them / writing new ones?

Comment (mgaeta) [reply tests/sentry/mail/test_adapter.py:219]: At first I removed the tests because I had removed the function but I just added everything back.

**Review by ceorourke — APPROVED** [review]

**Pull request merged**